### PR TITLE
graphql: return live objects from consistent object query

### PR DESCRIPTION
**Pull request by gegaowp**
## Description 

we plan to remove deleted / wrapped objects from objects_snapshot, this prep pr is to make the consistency query no longer return deleted / wrapped objects as a whole, after the removal.

## Test plan 

How did you test the new or updated feature?

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:
- [ ] REST API:

### CODE REVIEW

File: crates/sui-graphql-rpc/src/consistency.rs
Comment (gegaowp) [inline crates/sui-graphql-rpc/src/consistency.rs:170]: @wlmyng do you think we will need an index
>  (object_id, object_version, checkpoint_sequence_number, object_status)

instead of

>  (object_id, object_version, checkpoint_sequence_number)

Comment (wlmyng) [reply crates/sui-graphql-rpc/src/consistency.rs:170]: I think we would benefit more from having the existing indexes replaced with partial indexes, except for the index on `(object_id, object_version, checkpoint_sequence_number)`. This is because if we apply the filter on `object_status`, basically every index that supports a possible filter will also need to account for `object_status` in some way. Since `object_status` consists of 1 of 2 values, a partial index will do the job and comes with the added benefit of not indexing `WrappedOrDeleted` entries as well

The reason why I think we don't want a partial index on `(object_id, object_version, checkpoint_sequence_number)` is because of `objects_snapshot`:

```
explain select candidates.* from (select * from objects_snapshot and owner_id = '\x34f483dfc35589a1b78052d1b34ee5c6d8168a2dd982c54edf8fac7a0943a255') candidates left join (select object_id, object_version from objects_history where checkpoint_sequence_number between 30000 and 40000) newer on (candidates.object_id = newer.object_id and candidates.object_version < newer.object_version) where newer.object_version is null  order by candidates.object_id asc limit 50;
```

In this query, the planner leverages indexes on `objects_snapshot` filtering, and then uses the `object_id_object_version_checkp_idx` on `objects_history`, so that's the one index we don't actually want to modify

Comment (gegaowp) [reply crates/sui-graphql-rpc/src/consistency.rs:170]: makes sense!

> I think we would benefit more from having the existing indexes replaced with partial indexes, except for the index on (object_id, object_version, checkpoint_sequence_number)

however to replace it with partial indices, looks like PG does not allow me to do that in place. I will hold the indices change for now and do that after pruning is enabled so that can do that on only the last 2 partitions.

File: crates/sui-graphql-rpc/src/consistency.rs
Comment (gegaowp) [inline crates/sui-graphql-rpc/src/consistency.rs:179]: is this doing history object lookup, iiuc we will start pruning objects_history to 2 partitions, should we start reading this from `full_objects_history` instead.

Comment (wlmyng) [reply crates/sui-graphql-rpc/src/consistency.rs:179]: This is fine and desired because we only want to find yet newer versions of any objects that matched the filtering conditions (candidates)

File: crates/sui-graphql-rpc/src/consistency.rs
Comment (wlmyng) [reply crates/sui-graphql-rpc/src/consistency.rs:170]: That's smart, but I worry that objects queries will degrade in performance in the meantime ... but I can approve this to unblock and get the objects snapshot changes rolling (for removing wrapped and deleted)

Comment (wlmyng) [reply crates/sui-graphql-rpc/src/consistency.rs:170]: Could we create the partial indexes additively, and then drop the old indexes when they're ready

Comment (gegaowp) [reply crates/sui-graphql-rpc/src/consistency.rs:170]: makes sense, added a migration for that -- the clock is ticking after merging tho for the new indices to be ready, otherwise they will run on next release, the droppings are quick, hopefully we can enable pruning soon and can only create new partial indices for the last 2 partiions.

File: crates/sui-graphql-rpc/src/types/object.rs
Comment (amnn) [inline crates/sui-graphql-rpc/src/types/object.rs:753]: Do we ever pass `None` to this now?

File: crates/sui-indexer/migrations/pg/2024-10-08-025030_partial_index_instead/down.sql
Comment (amnn) [inline crates/sui-indexer/migrations/pg/2024-10-08-025030_partial_index_instead/down.sql:1]: When we create these indices, we should use different names for the new partial indices, so that they can co-exist with the existing indices. This is because it will take quite some time to build these indices, and we don't want the object tables to be without any indices in that time (we should run the index creation statements proactively).

Also, how long does it take to drop a large index, usually? If it can take some time, we would want  to run those proactively too, and in that case we should split them up into their own migration.

**Review by amnn — COMMENTED** [review]
Had some questions about the migrations for building indices, but otherwise this change looks good, thanks @gegaowp .

File: crates/sui-indexer/migrations/pg/2024-10-08-025030_partial_index_instead/down.sql
Comment (gegaowp) [reply crates/sui-indexer/migrations/pg/2024-10-08-025030_partial_index_instead/down.sql:1]: > we should use different names for the new partial indices

good point.

> how long does it take to drop a large index

it's mostly within a min from my previous index droppings -- here I did plan to do both drop and creations proactively like you did so that when the actual migration runs, it will check `IF (NOT) EXISTS` and basically do no-op.

File: crates/sui-graphql-rpc/src/types/object.rs
Comment (wlmyng) [reply crates/sui-graphql-rpc/src/types/object.rs:753]: Yes I think this can be non optional now, we'll always have bcs

**Review by wlmyng — APPROVED** [review]

File: crates/sui-indexer/migrations/pg/2024-10-08-025030_partial_index_instead/down.sql
Comment (amnn) [reply crates/sui-indexer/migrations/pg/2024-10-08-025030_partial_index_instead/down.sql:1]: The more I think about it, the DROPs need to be in their own migration, that you don't land with this PR. The release needs to be staged as follows:

- Create indices proactively.
- Release new indexer with migration to confirm those indices exist.
- Release new GraphQL that uses the partial indices -- confirm they are working.
- Drop the old indices proactively.
- Land the migration that "officially" removes the indices.

Otherwise, there will be a period of time where the old GraphQL service is running and doesn't have its indices available. And if there is some issue with the new indices this could be measured in days.

**Comment by Emadalshamery** [comment]
I'm ready

في الأربعاء، 9 أكتوبر 2024, 12:47 م Ashok Menon ***@***.***>
كتب:

File: crates/sui-indexer/migrations/pg/2024-10-08-025030_partial_index_instead/down.sql
Comment (gegaowp) [reply crates/sui-indexer/migrations/pg/2024-10-08-025030_partial_index_instead/down.sql:1]: I think that's fair, will split that then.

File: crates/sui-graphql-e2e-tests/tests/stable/call/dynamic_fields.exp
Comment (amnn) [inline crates/sui-graphql-e2e-tests/tests/stable/call/dynamic_fields.exp:1]: @gegaowp -- this test change seems not right.  Why would we expect this change to cause dynamic field queries to start returning `null`?

File: crates/sui-graphql-e2e-tests/tests/stable/consistency/dynamic_fields/immutable_dof.exp
Comment (amnn) [inline crates/sui-graphql-e2e-tests/tests/stable/consistency/dynamic_fields/immutable_dof.exp:147]: Likewise -- this seems unexpected.

**Review by amnn — COMMENTED** [review]
Some test changes here seem unexpected, PTAL.

File: crates/sui-graphql-e2e-tests/tests/stable/call/dynamic_fields.exp
Comment (amnn) [reply crates/sui-graphql-e2e-tests/tests/stable/call/dynamic_fields.exp:1]: Took a look -- the issue is that the parent was wrapped in this case, and previously we supported still querying for the parent and then fetching its dynamic fields but now we don't, because we don't return a result for a wrapped object.

Best thing to do here would be to adapt this test to highlight the difference. I can do that.

File: crates/sui-graphql-e2e-tests/tests/stable/consistency/dynamic_fields/immutable_dof.exp
Comment (amnn) [reply crates/sui-graphql-e2e-tests/tests/stable/consistency/dynamic_fields/immutable_dof.exp:147]: This one is a bit more nuanced -- the parent is a wrapped object, and because of this change, we can now no longer get its ID. This will require further schema changes to accommodate, I think.

File: crates/sui-graphql-e2e-tests/tests/stable/call/dynamic_fields.exp
Comment (gegaowp) [reply crates/sui-graphql-e2e-tests/tests/stable/call/dynamic_fields.exp:1]: thanks for the double-check, this one and the other were not correct to me at the first glance either -- good idea that we want to test them more explicitly, thanks!

File: crates/sui-graphql-e2e-tests/tests/stable/consistency/dynamic_fields/immutable_dof.exp
Comment (gegaowp) [reply crates/sui-graphql-e2e-tests/tests/stable/consistency/dynamic_fields/immutable_dof.exp:147]: afaik we overall lack support for wrapped objects like their type, children etc. today, it can be a sizable change across many RPC surfaces if we want to be consistent.

**Comment by amnn** [comment]
https://github.com/MystenLabs/sui/pull/19785 addresses the comments I made around tests.

**Pull request merged**
